### ISSUE TITLE
-Wunsafe-buffer-usage erroneously warns on array accesses into compile-time string constants, but only in function scope

### ISSUE THREAD

**Issue by pkasting** [llvm/llvm-project#92191]
Compiling the following code with `-Wunsafe-buffer-usage` unexpectedly emits a warning:
```
constexpr char kAbc[] = "abc";
static_assert(kAbc[3] == '\0');  // OK

static_assert("abc"[3] == '\0');  // OK

void f() {
  static_assert(kAbc[3] == '\0');  // OK

  static_assert("abc"[3] == '\0');  // WARNING
}
```
Live repro: https://godbolt.org/z/c6jhdvPTb

**Comment by pkasting** [comment]
Here's a different, probably-related testcase:
```
constexpr char kBuf[64] = {};

constexpr const char* p = kBuf + 1;  // OK

void f() {
  static constexpr const char* p = kBuf + 1;  // WARNING
}
```
Live repro: https://godbolt.org/z/7T6edKMz8

**Comment by pkasting** [comment]
@haoNoQ @jkorous-apple This is similar to issue 64646, but less defensible, since here we are not passing through a function parameter and the size truly is visible.

I realize that https://github.com/llvm/llvm-project/issues/64646#issuecomment-1749309422 is probably addressing this bug directly. It would be good to know the current status/future timeline of this warning, as Chromium devs have already turned it on by default (but blocklisted most of the codebase) and are starting to try to unblock it globally. If it's "known to be glaringly unready for production", that's probably valuable input for their planning...

**Comment by jkorous-apple** [comment]
Thank you for the examples!
We originally went with a very simple and robust approach that ANY pointer arithmetic is unsafe.
These cases might warrant us rethinking that.

**Comment by jkorous-apple** [comment]
Actually, we don't even have to rethink anything since we already allow these for subscripts.
https://github.com/llvm/llvm-project/pull/80504

It only makes sense to keep the concept of unsafe arithmetic consistent with unsafe subscript operation.

**Comment by haoNoQ** [comment]
Yeah there are a few things going on here:
1. We miss the part where `"abc"[3]` is safe because the buffer size is known to be large enough at compile time.
2. We miss the part where `static_assert` conditions aren't actually executed, so in fact this code cannot lead to an out-of-bounds access regardless of the array size or index value.
3. Finally, it looks like we don't analyze the top-level code at all, which explains why there's no warning outside the function.

At a glance I think the order of importance is 3 > 2 > 1 but I'm very open to changing my mind.

**Comment by jkorous-apple** [comment]
I take this issue is about 1 and will try to address that soon.
Let's split off 2 and 3 as separate issues.

**Comment by pkasting** [comment]
> Let's split off 2 and 3 as separate issues.

Sounds good; do you plan to do that or do you want me to take a shot?

**Comment by jkorous-apple** [comment]
@pkasting go for it.

**Comment by pkasting** [comment]
I filed issue 92664 and issue 92665 for points 3 and 2, respectively.
